rust-peer/src/peer.rs

```diff
@@ -0,0 +1,522 @@
+            let multiaddr: Multiaddr = peer.parse().expect("Failed to parse Multiaddr");
+            if let Err(e) = self.swarm.dial(multiaddr) {
+                debug!("Failed to dial {peer}: {e}");
+            }
+        }
+
+        // Initialize the gossipsub topics and hashes
+        let chat_topic = GossipsubIdentTopic::new(GOSSIPSUB_CHAT_TOPIC);
+        let file_topic = GossipsubIdentTopic::new(GOSSIPSUB_CHAT_FILE_TOPIC);
+        let peer_discovery = GossipsubIdentTopic::new(GOSSIPSUB_PEER_DISCOVERY);
+        let chat_topic_hash = chat_topic.hash();
+        let file_topic_hash = file_topic.hash();
+        let peer_discovery_hash = peer_discovery.hash();
+
+        // Subscribe to the gossipsub topics
+        for topic in &[chat_topic, file_topic, peer_discovery] {
+            if let Err(e) = self.swarm.behaviour_mut().gossipsub.subscribe(topic) {
+                debug!("Failed to subscribe to topic {topic}: {e}");
+            }
+        }
+
+        // Create our loop ticker
+        let mut tick = tokio::time::interval(Duration::from_millis(18));
+
+        // Run the main loop
+        loop {
+            // process messages from the UI
+            if let Ok(message) = self.from_ui.try_recv() {
+                match message {
+                    Message::Chat { data, .. } => {
+                        error!("chat received");
+                        if let Err(e) = self
+                            .swarm
+                            .behaviour_mut()
+                            .gossipsub
+                            .publish(chat_topic_hash.clone(), data)
+                        {
+                            debug!("Failed to publish chat message: {e}");
+                        }
+                    }
+                    Message::AllPeers { .. } => {
+                        error!("all peers received");
+                        let peers = self
+                            .swarm
+                            .behaviour()
+                            .gossipsub
+                            .all_peers()
+                            .filter(|(_, topics)| !topics.is_empty())
+                            .map(|(peer_id, topics)| {
+                                (*peer_id, topics.iter().map(|t| t.to_string()).collect())
+                            })
+                            .collect();
+                        self.to_ui.send(Message::AllPeers { peers }).await?;
+                    }
+                    _ => {
+                        debug!("Unhandled message: {:?}", message);
+                    }
+                }
+            }
+
+            tokio::select! {
+                _ = self.shutdown.cancelled() => {
+                    info!("Shutting down the peer");
+                    break;
+                }
+
+                _ = tick.tick() => {}
+
+                Some(event) = self.swarm.next() => match event {
+
+                    // When we figure out what our external address is
```

Comment:
Why can we assume that the listening address is an external address? 
If we are using the default listen ip `0.0.0.0` that the transports will listen on all interfaces, so the address reported here might be e.g. a loopback address.

Reply:
Ah, good catch.